pep-0712.rst

```diff
@@ -0,0 +1,326 @@
+library means more users are able to opt-in to these benefits without requiring
+third-party libraries. Additionally third-party libraries are able to clue
+type-checkers into their own conversion semantics through added support in
+:func:`~py3.11:typing.dataclass_transform`, meaning users of those libraries
+benefit as well.
+
+Specification
+=============
+
+New ``converter`` parameter
+---------------------------
+
+This specification introduces a new parameter named ``converter`` to the
+:func:`dataclasses.field` function. When an ``__init__`` method is synthesized
+by ``dataclass``-like semantics, if an argument is provided for the field, the
+``dataclass`` object's attribute will be assigned the result of calling the
+converter on the provided argument. If no argument is given and the field was
+constructed with a default value, the ``dataclass`` object's attribute will be
+assigned the result of calling the converter on the provided default.
+
+Adding this parameter also implies the following changes:
+
+* A ``converter`` attribute will be added to :class:`dataclasses.Field`.
+* ``converter`` will be added to the field specifier parameters of arguments
+  provided to :func:`~py3.11:typing.dataclass_transform`'s ``field`` parameter.
+
+Example
+'''''''
+
+.. code-block:: python
+
+  @dataclasses.dataclass
+  class InventoryItem:
+      # `converter` as a type
+      id: int = dataclasses.field(converter=int)
+      skus: tuple[int, ...] = dataclasses.field(converter=tuple[int, ...])
+      # `converter` as a callable
+      names: tuple[str, ...] = dataclasses.field(
+        converter=lambda names: tuple(map(str.lower, names))
+      )
+
+      # The default value is also converted; therefore the following is not a
+      # type error.
+      stock_image_path: pathlib.PurePosixPath = dataclasses.field(
+        converter=pathlib.PurePosixPath, default="assets/unknown.png"
+      )
+
+  item1 = InventoryItem("1", [234, 765], ["PYTHON PLUSHIE", "FLUFFY SNAKE"])
+  # item1 would have the following values:
+  #   id=1
+  #   skus=(234, 765)
+  #   names=('python plushie', 'fluffy snake')
+  #   stock_image_path=pathlib.PurePosixPath("assets/unknown.png")
+
+
+Impact on typing
+----------------
+
+``converter`` arguments are expected to be callable objects which accept a
+single argument and return a type compatible with the field's annotated type.
+The type of the callable's argument is used as the type of the corresponding
+parameter in the synthesized ``__init__`` method.
+
+Type-narrowing the argument type
+''''''''''''''''''''''''''''''''
+
+For the purpose of deducing the type of the argument in the synthesized
+``__init__`` method, the ``converter`` argument's type can be "narrowed" using
+the following rules:
+
+* If the ``converter`` is of type ``Any``, it is assumed to be callable with a
+  single ``Any`` typed-argument.
+* All keyword-only parameters can be ignored.
+* ``**kwargs`` with defaults can be ignored.
```

Comment:
I find this entire section to be very confusing the way it is worded.

I think what you are trying to say is:
A converter must be a callable that accepts a single positional argument, and the parameter type corresponding to this positional argument provides the type of the the synthesized `__init__` parameter associated with the field. In other words, the argument provided for the `converter` parameter must be compatible with `Callable[[T], X]` where `T` is the input type for the converter and `X` is the output type of the converter.

Does that cover it? If so, then I recommend removing all of the additional text about default parameters, `*args`, `**kwargs`, etc. because they are irrelevant and will lead to confusion. All of those rules (and others) fall out naturally from the fact that the type must be compatible with `Callable[[T], X]`.

Reply:
Yup, that's much easier to type and read.

Sometimes less if more 😅